odinmain/src/main/kotlin/me/odinmain/features/impl/floor7/DragonCheck.kt

```diff
@@ -14,55 +11,36 @@ import me.odinmain.features.impl.floor7.WitherDragons.sendTime
+import net.minecraft.network.play.server.S0FPacketSpawnMob
+import net.minecraft.network.play.server.S1CPacketEntityMetadata
+import net.minecraft.util.Vec3
 import net.minecraftforge.event.entity.EntityJoinWorldEvent
-import net.minecraftforge.event.entity.living.LivingDeathEvent
 
 object DragonCheck {
 
     var lastDragonDeath: WitherDragonsEnum = WitherDragonsEnum.None
-    var dragonEntityList = emptyList<EntityDragon>()
+    val dragonEntityList = mutableListOf<EntityDragon>()
 
-    fun dragonJoinWorld(event: EntityJoinWorldEvent) {
-        val entity = event.entity as? EntityDragon ?: return
-        val dragon = WitherDragonsEnum.entries.find { isVecInXZ(entity.positionVector, it.boxesDimensions) } ?: return
-        if (dragon.state == WitherDragonState.DEAD) return
-
-        dragon.state = WitherDragonState.ALIVE
-        dragon.timeToSpawn = 100
-        dragon.timesSpawned += 1
-        dragon.entity = entity
-        dragon.spawnedTime = System.currentTimeMillis()
-        dragon.isSprayed = false
-
-        if (sendArrowHit && WitherDragons.enabled) arrowSpawn(dragon)
-        if (resetOnDragons && WitherDragons.enabled) onDragonSpawn()
-        if (sendSpawned && WitherDragons.enabled) {
-            val numberSuffix = when (dragon.timesSpawned) {
-                1 -> "st"
-                2 -> "nd"
-                3 -> "rd"
-                else -> "th"
-            }
-            modMessage("§${dragon.colorCode}${dragon.name} §fdragon spawned. This is the §${dragon.colorCode}${dragon.timesSpawned}${numberSuffix}§f time it has spawned.")
-        }
+    fun dragonUpdate(packet: S1CPacketEntityMetadata) {
+        val dragon = WitherDragonsEnum.entries.find { it.entityId == packet.entityId } ?: return
+        if (dragon.entity == null) return dragon.updateEntity(packet.entityId)
+        val health = packet.func_149376_c().find { it.dataValueId == 6 }?.`object` as? Float ?: return
+        if (health > 0 || dragon.state == WitherDragonState.DEAD) return
+        dragon.setDead()
     }
 
-    fun dragonLeaveWorld(event: LivingDeathEvent) {
-        val entity = event.entity as? EntityDragon ?: return
-        val dragon = WitherDragonsEnum.entries.find { it.entity?.entityId == entity.entityId } ?: return
-        dragon.state = WitherDragonState.DEAD
-        lastDragonDeath = dragon
-
-        if (sendTime && WitherDragons.enabled)
-            dragonPBs.time(dragon.ordinal, entity.ticksExisted / 20.0, "s§7!", "§${dragon.colorCode}${dragon.name} §7was alive for §6", addPBString = true, addOldPBString = true)
-
-        if (sendArrowHit && WitherDragons.enabled) arrowDeath(dragon)
+    fun dragonSpawn(packet: S0FPacketSpawnMob) {
+        val dragon = WitherDragonsEnum.entries
+            .find { isVecInXZ(Vec3(packet.x / 32.0, packet.y / 32.0, packet.z / 32.0), it.boxesDimensions) }
+            ?.takeIf { it.state != WitherDragonState.SPAWNING } ?: return
+        dragon.setAlive(packet.entityID)
```

Comment:
Dragon does not need to be a variable since it is only used once, do the findinging of the entity and run setAlive directly on the result.